CONTRIBUTING.md

```diff
@@ -0,0 +1,147 @@
+        - `out[0, 0, 0]` or `out[idx]/out[curr]/out[something]`
+
+Flake8 rules not listed will be enforced unless we find a need not to enforce them.
+Since documentation does go out of date, please consult the entry in `setup.cfg` for the most up-to-date requirements and update this document if you notice a difference.
+
+## Style
+
+The first version of the dycore was written with minimal metadata and typing, motivated primarily by matching the regression data produced by the Fortran version of the code using the numpy backend.
+We are now actively refactoring, moving code that still does computations in Python into GT4py stencils and merging stencils together with the introduction of enabling features in GT4py (such as regions).
+While we do that, clarifying the operation of the model and what the variables are will both help make the model easier to read and reduce errors as we move around long lists of argument variables.
+
+Specifically, we want to start adding the following where appropriate:
+- Type hints on Python functions (see [`utils/typing.py`](https://github.com/VulcanClimateModeling/fv3core/blob/master/utils/typing.py) and below)
+- More descriptive types on stencil definitions
+- Docstrings on outward facing Python functions: describe what methods are doing, describe the intent (*in*, *out*, or *inout*) of the function arguments
+
+### Docstrings
+These should aid us in refactoring and understanding what a function is doing. If it is not completely understood yet what is happening, what is known can be written along with a `TODOC` to indicate it is incomplete.
+For example:
+
+```python
+def stencil(...):
+    """This is a short description that fits on one line.
+
+    Here is a longer explanation of the assumptions and why this exists.
+
+    TODOC: I do not fully understand why ke_c needs to be updated here.
+
+    Args:
+        uc: x-velocity on C-grid (inout)
+        vc: y-velocity on C-grid (inout)
+        vort_c: Vorticity on C-grid (inout)
+        ke_c: kinetic energy on C-grid (inout)
+        v: y-velocity on D-grid (inout)
+        u: x-velocity on D-grid (inout)
+        dt2: timestep (in)
+
+    """
+```
+
+
+### Python functions
+These should mostly be lightweight workflow wrappers calling gt4py stencils, though currently exceptions exist where Python code does computations on data fields.
+
+Original convention is:
+```python
+def compute(var1, var2, var3, param1, param2, param3):
+```
+
+Order of arguments does not actually matter, but generally follows the convention of listing 3d fields first, followed by parameters, as is required by gt4py stencil functions.
+
+New convention: make use of `fv3core/utils/typing.py` to specify fields, also type-hint any function
+outputs.
+
+For example:
+```python
+def compute(var1: FloatField, var2:IntField, var3: BoolField,
+            param1: float_type, param2: int_type, param3: bool_type):
+```
```

Comment:
Because of how github messages/formatting work, I had to put backslashes before the code blocks - please remove them from my suggestion if/when you merge it.

```suggestion
New code should be type hinted making use of `fv3core/utils/typing.py` when typing gt4py fields. You may run into code in fv3core before we added this convention. An older code like:
\```python
def compute(var1, var2, var3, param1, param2, param3):
\```

would become:

\```python
def compute(var1: FloatField, var2:IntField, var3: BoolField,
            param1: float_type, param2: int_type, param3: bool_type):
\```
There is no determined convention for order of arguments, but the code generally follows the convention of listing 3d fields first followed by parameters, as is required by gt4py stencil functions.
```

Reply:
Suggestion still outstanding.